### PR TITLE
Pathname of anchor tag with relative `href`

### PR DESCRIPTION
**Pull request by Sebmaster**
I was working with JSDom and ran into some behavior with `myAnchorTag.pathname` that I thought was unexpected. Let me show an example:

```js
global.jsdom = require('jsdom').jsdom;
global.document = jsdom('<html><head><script></script></head><body></body></html>');
global.window = global.document.parentWindow;
global.navigator = window.navigator = { userAgent: 'JSDom', appVersion: '' };

var a = document.createElement('a');

a.href = "uri/what/pls";
a.pathname;
// /what/pls

a.href = "/uri/what/pls";
a.pathname;
// /uri/what/pls
```

Is it unexpected that the pathname always strips the first part of the path when it is relative + no leading slash? Or am I missing something obvious here? :open_mouth:

### CODE REVIEW

**Comment by Sebmaster** [comment]
Appended a PR here which fixes this issue and should pass all tests to get the discussion here started.

I'm still not completely sure about urlmaster, I'll maybe experiment a bit with different url parsers, maybe [urlparser](https://github.com/petkaantonov/urlparser) (as per @domenic's [comment](https://github.com/tmpvar/jsdom/pull/1037#issuecomment-75272058)

**Comment by Sebmaster** [comment]
Okay I switched to URIjs again. There's an issue there with the percent encoding which I reported there (medialize/URI.js#198), but I think it's the most standard conformant one (although it seems we still have to handle our about:blanks manually).

I think the only other failing test is one where the expect cases are wrong, which are actually more right with URIjs. The test cases are auto generated by urlmaster, however it seems it doesn't actually resolve the testcases itself well enough. We'll need to either convert them to explicit tests or remove them

File: lib/jsdom.js [hunk 1]
```diff
@@ -361,6 +364,12 @@ function getConfigFromArguments(args, callback) {
     config.url = toFileUrl(config.file);
   }
 
+  if (config.url) {
+    // backslashes were originally not allowed in URLs, so this is error-handling only
+    // comparable to chrome which converts all backslashes to foward slashes
+    config.url = config.url.replace(/\\/g, '/');
```
Comment (domenic) [inline lib/jsdom.js]: Hmm shouldn't we just do config.url = url.resolve("about:blank", config.url)` which would normalize everything for us?

**Comment by Sebmaster** [comment]
We should add the [URL tests](https://github.com/w3c/web-platform-tests/tree/master/url) from web-platform-tests too here.

**Comment by Sebmaster** [comment]
Oooookay, I got some time to fix this up again. A few things I still know which are left (there might be a few more):

- [ ] Try to completely throw away `fixFileUrl` again so we don't break back-compat (also because I don't know if (or where) it's actually necessary)
- [ ] Run tests, fix them (remove urlmaster?)
- [ ] Either merge the two spec/* branches in jsdom/whatwg-url into master (which would make master deviate from the spec), try to get it into the spec somewhat quickly (or something else?).
- [ ] Check into the comments on [jsdom/whatwg-url#e8b2dbe](https://github.com/jsdom/whatwg-url/commit/e8b2dbefc36335fa9db6d689f424ec2eb2ca92d9).

**Comment by Sebmaster** [comment]
- [ ] `removeAttribute` should reset the URL object somehow apparently (does in Chrome at least); `href` returns "" after.

**Comment by Sebmaster** [comment]
Oh buy, fixing these tests is not fun. Windows fails because some dependency issues between tests (a previous tests kills the level2/style test), Linux/Travis fails on the filename_with_spaces test :(

**Comment by domenic** [comment]
I'll take care of it :)

**Comment by Sebmaster** [comment]
We have passing tests! @domenic please review.

**Comment by domenic** [comment]
I am ready to merge this as soon as you give me the sign off btw. The nits and whatwg-url issues can be fixed in follow-ups or as I merge. Just want to make sure you're happy with it first.

**Comment by Sebmaster** [comment]
Should be good to go for now.

**Comment by domenic** [comment]
This hangs locally on w3c/index.js... going to try to track it down.

**Comment by Sebmaster** [comment]
Windows? Might be our classic path resolving then. Might end up putting the patch in compat too...

**Comment by domenic** [comment]
I bet it's the usage of things like .href and such

**Comment by domenic** [comment]
Yeah resource.url.href is `c:/Users/Domenic/Dropbox/GitHub/tmpvar/jsdom/test/w3c/tests/resources/testharness.js` missing `file:///` and capitalization wrong. Working on a fix.

**Comment by jmeas** [comment]
:tada:

**Comment by domenic** [comment]
v6.0.0 coming up soon...

**Pull request merged**